volumetric/volMask.cpp

```diff
@@ -0,0 +1,255 @@
+///////////////////////////////////////////////////////////////////////////////
+using namespace std;
+using namespace DGtal;
+///////////////////////////////////////////////////////////////////////////////
+namespace po = boost::program_options;
+
+
+/**
+ @page volMask volMask
+ 
+ @brief  Extracts a new image from the a mask image that represents the regions of the image which are selected and copied in the resulting image. Elements outside the regions defined by the mask are set to 0.
+ 
+ @b Usage:   volMask [input]
+ 
+ @b Allowed @b options @b are :
+ 
+ @code
+ -h [ --help ]               display this message
+ -i [ --input ] arg          an input vol (or ITK: .nii, mha, ... ) file.
+ -t [ --inputType ] arg      to sepcify the input image type (int or double).
+ -a [ --mask ] arg           the mask image that represents the elements that
+                             are copied as output in the resulting image (by
+                             default set to 1 you can change this value by
+                             using --maskValue).
+ -o [ --output ] arg         the output masked image.
+ -m [ --maskValue ] arg (=1) the masking value.
+ @endcode
+ 
+ @b Example:
+ 
+ @code
+ volMask -i ${DGtal}/examples/samples/lobster.vol  -a ${DGtal}/examples/samples/lobster.vol -o lobsMasked.vol -m 100
+ @endcode
+ 
+ @image html resvolMask.png "Example of result. "
+ 
+ @see
+ @ref volMask.cpp
+ 
+ */
+
+
+
+
+typedef ImageContainerBySTLVector < Z3i::Domain, unsigned char > Image3D;
+#ifdef WITH_ITK
+typedef ImageContainerBySTLVector < Z3i::Domain,  double > Image3D_D;
+typedef ImageContainerBySTLVector < Z3i::Domain,  int > Image3D_I;
+#endif
+
+
+
+template<typename TImage, typename TImageMask>
+typename TImageMask::Domain
+subDomainMasked(const TImage &image, const  TImageMask &maskImage,
+                typename TImageMask::Value maskValue){
+  typename TImageMask::Domain res;
+  Z3i::Point minP = image.domain().upperBound();
+  Z3i::Point maxP = image.domain().lowerBound();
+  
+  Z3i::Point::Iterator minIt;
+  Z3i::Point::Iterator maxIt;
+  
+  for(const auto &p: image.domain())
+  {
+    minIt = minP.begin();
+    maxIt = maxP.begin();
+    maxIt = maxP.begin();
+    if( maskImage.domain().isInside(p) && maskImage(p) ==  maskValue ) // no noise on mask image
+    {
+      for(auto pIt=p.begin(); pIt!=p.end();pIt++ )
+      {
+        if( *pIt < *minIt ){*minIt = *pIt;}
+        if( *pIt > *maxIt ){*maxIt = *pIt;}
+        minIt++;
+        maxIt++;
+      }
+    }
+  }
+  
+  // offset to avoid problems on borders
+  Z3i::Point offset(5,5,5);
+  minP -= offset;
+  maxP += offset;
+  trace.info() << "sub-domain:" << minP << " " << maxP << std::endl;
+  return typename TImageMask::Domain(minP, maxP);
+}
+
+
+template<typename TImage, typename TImageMask>
+void
+applyMask(const TImage &inputImage,TImage &outputImage,
+          const TImageMask &maskImage, typename TImageMask::Value maskValue)
+{
+  for (const auto &p: outputImage.domain())
```

Comment:
So because of the 5 offset on the border, you might get a bigger image than input image (and that is why you must check that your point is inside the input domain) ?

Reply:
no the check is because in some cases the mask image do not have necessary the same domain of the input image.